### ISSUE TITLE
How to access LruCache<&'a str, String> with the key of String?

### ISSUE THREAD

**Issue by wfxr** [marmeladema/clru-rs#25]
It looks like `get` requires the argument has the same lifetime as it's key.

```rust
use clru::CLruCache;
use std::collections::HashMap;

struct Store<'a> {
    index: HashMap<&'a str, String>,
    cache: CLruCache<&'a str, String>,
}

impl<'a> Store<'a> {
    fn foo(&mut self, key: String) -> String {
        self.index.get(&key.as_str()).unwrap().to_owned()
    }

    fn bar(&mut self, key: String) -> String {
        self.cache.get(&key.as_str()).unwrap().to_owned()
    }
}
```

How can I get the method `bar` to compile?

*I firstly opend a same issue at [lru-rs](https://github.com/jeromefroe/lru-rs). Then I found this crate and tried it. But it also doesn't compile. Maybe I did not use it in a correct way?*

**Comment by marmeladema** [comment]
Hello @wfxr and thank you for taking a look at this crate.

The code you wrote is *almost* correct, you just added a `&` to take a reference of the reference returned by `String::as_str` which makes the borrow checker complain.

If you do that instead, it should work just fine:
```
use clru::CLruCache;
use std::collections::HashMap;

struct Store<'a> {
    index: HashMap<&'a str, String>,
    cache: CLruCache<&'a str, String>,
}

impl<'a> Store<'a> {
    fn foo(&mut self, key: String) -> String {
        self.index.get(key.as_str()).unwrap().to_owned()
    }

    fn bar(&mut self, key: String) -> String {
        self.cache.get(key.as_str()).unwrap().to_owned()
    }
}
```

Additionally, if you are creating collections with borrowed string as key, you should not need to pass owned string for `get` (and all *read* related functions) and just use a plain `key: &str` instead. This way you avoid having to systematically requires that the caller allocates memory.

**Comment by wfxr** [comment]
@marmeladema Thank you very much! I never relized I need to remove the `&`. The reason for using `String` instead of `&str` is that I actually need to implement a trait whose signature requires a parameter of `String` and I need to use the cache inner it. That's why I wrote these weird codes :)

**Comment by marmeladema** [comment]
For the full story, `String` implements:
* `Deref<Target = str>`: https://doc.rust-lang.org/std/string/struct.String.html#deref-methods which means you can also write: `&*key` (or even `key.deref()` if you import the `std::ops::Deref` trait in scope)
* `Borrow<str>`: https://doc.rust-lang.org/std/string/struct.String.html#impl-Borrow%3Cstr%3E which means you can also write: `&key.borrow()`

Both are equivalent to `String::as_str` and will return a borrowed version of the owned `String` object: `&'a String -> &'a str`.
By creating another reference you are actually ending up with a `&'_ &'a str` which is a reference to a reference. And as borrowing rules dictate the first reference needs to live _at least as long as_ the second one (ie: `'a: '_`). In your case you are not keeping the second reference longer than necessary but the borrow checking gets confused. There are probably good reasons it behaves like that but clearly it can be confusing on contrived examples like the one your shared.

**Comment by wfxr** [comment]
Thank you for such a detailed explanation, I learned a lot from it.

I also wonder why `Hashmap`'s `get` accepts the extra `&`? ( Comment out the `bar` method and the initial demo will compile. )

**EDIT**:

After look into each `get`'s implementation I found `get` in `Hashmap` doesn't have a generic parameter of lifetime but `CLruCache` does have. I think I need to read more sources to learn the knowledge about the life time.
